### PR TITLE
Softmax variant refactor

### PR DESCRIPTION
**Pull request by gkielian**
Regularized the parameter inputs for each of the softmax variant nn.Modules to simplify addition of new softmax style variations.

This change makes model.py a lot cleaner and simplifies the exploration scripts for scanning across the softmax variations (no more boolean values, now one only has to loop over bash string arrays).

I think there will be merge conflicts with other PRs, so I'm planning on merge these change after the current PRs.

### CODE REVIEW

**Comment by gkielian** [comment]
This was a relatively large overhaul.

## Changes:
- using 'config' file as central means to setup nn.Modules -- this is the main change, there may now be a way to leverage polymorphism to make the file a lot cleaner, but definitely feel this is the first step.
- New specific settings for
  - softmax of the attention layers
  - softmax of the output layer (defaults to regular softmax)
  - activation types
  - normalization types
 
## Formatting:
- added an import for python pretty printing

## Tests:
- added a `tests` directory intended to hold tests for our git runners -- we can adapt our explorations
- as a first test, adding a "test all softmaxes" test, doing both training and inference success verification

**Review by klei22 — APPROVED** [review]
Tested main variations, looks good.
